Source/ROUtils/Utils/DragCubeTool.cs

```diff
@@ -0,0 +1,174 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Profiling;
+
+namespace ROUtils
+{
+    public class DragCubeTool : MonoBehaviour
+    {
+        private static readonly Dictionary<string, DragCube> _cacheDict = new Dictionary<string, DragCube>();
+        private static bool _statsRoutineStarted = false;
+        private static uint _cubesRenderedThisFrame = 0;
+        private static uint _cacheHitsThisFrame = 0;
+        private static long _elapsedTicks = 0;
+
+        private string _shapeKey;
+
+        public static bool UseCache { get; set; } = true;
+        public static bool ValidateCubes { get; set; } = false;
+        public static uint MaxCacheSize { get; set; } = 5000;
+
+        public Part Part { get; private set; }
+
+        public static DragCubeTool UpdateDragCubes(Part p, string shapeKey = null)
+        {
+            var tool = p.GetComponent<DragCubeTool>();
+            if (tool == null)
+            {
+                tool = p.gameObject.AddComponent<DragCubeTool>();
+                tool.Part = p;
+                tool._shapeKey = shapeKey;
+            }
+            return tool;
+        }
+
+        public static void UpdateDragCubesImmediate(Part p, string shapeKey = null)
+        {
+            if (!Ready(p))
+                throw new InvalidOperationException("Not ready for drag cube rendering yet");
+
+            UpdateCubes(p, shapeKey);
+        }
+
+        public void FixedUpdate()
+        {
+            if (Ready())
+                UpdateCubes();
+        }
+
+        public bool Ready() => Ready(Part);
+
+        private static bool Ready(Part p)
+        {
+            if (HighLogic.LoadedSceneIsFlight)
+                return FlightGlobals.ready;
+            if (HighLogic.LoadedSceneIsEditor)
+                return p.localRoot == EditorLogic.RootPart && p.gameObject.layer != LayerMask.NameToLayer("TransparentFX");
+            return true;
+        }
+
+        private void UpdateCubes()
+        {
+            UpdateCubes(Part, _shapeKey);
+            Destroy(this);
+        }
+
+        private static void UpdateCubes(Part p, string shapeKey = null)
+        {
+            if (ModUtils.IsFARInstalled)
+                p.SendMessage("GeometryPartModuleRebuildMeshData");
+
+            Profiler.BeginSample("UpdateCubes");
+            var sw = System.Diagnostics.Stopwatch.StartNew();
```

Comment:
Use a static one and call the relevant methods to do the accounting: https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch?view=netframework-4.7.2#methods